File: C++/Problems/Problems/problem_3804.cpp
#include "problem_3804.h"

#include <unordered_set>

bool problem_3804::test() {
    std::vector nums  { -1, 1, 0 };

    int expected = 5;

    auto result = centeredSubarrays(nums);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_3804::centeredSubarrays(std::vector<int> &nums) {
    int n = nums.size();
    int result = 0;
    for (int i = 0; i < n; ++i)
    {
        int sum = 0;
        std::unordered_set<int> present;
        for (int j = i; j < n; ++j)
        {
            sum += nums[j];
            present.insert(nums[j]);
            if (present.contains(sum))
                result++;
        }
    }

    return result;
}

File: Python/Problems/Problems/problem3804.py
import heapq
import math
import sys

import numpy as np

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem

class Problem3804(Problem):
    def test(self) -> bool:
        nums = [-1, 1, 0]

        expected = 5

        result = self.centeredSubarrays(nums)

        print(result)

        return result == expected

    def centeredSubarrays(self, nums: List[int]) -> int:
        n = len(nums)
        result = 0
        for i in range(n):
            sum = 0
            present = set()
            for j in range(i, n):
                sum += nums[j]
                present.add(nums[j])
                if sum in present:
                    result += 1

        return result